pkg/generator/patch.go

```diff
@@ -0,0 +1,144 @@
+package generator
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+	"sigs.k8s.io/kustomize/kyaml/yaml/merge2"
+	sigyaml "sigs.k8s.io/yaml"
+)
+
+// Apply strategic merge or json patch to the source YAML.
+func (gen *AssetGenerator) applyAssetPatch(sourceYAML []byte, patchAssetName string, extraReplacements []string) ([]byte, error) {
+	if strings.HasSuffix(patchAssetName, ".patch") {
+		return gen.applyJSONPatch(sourceYAML, patchAssetName, extraReplacements)
+	}
+	return gen.applyStrategicMergePatch(sourceYAML, patchAssetName, extraReplacements)
+}
+
+// Apply strategic merge patch to the source YAML.
+func (gen *AssetGenerator) applyStrategicMergePatch(sourceYAML []byte, patchAssetName string, extraReplacements []string) ([]byte, error) {
+	patchYAML := gen.mustReadAsset(patchAssetName, extraReplacements)
+	opts := yaml.MergeOptions{ListIncreaseDirection: yaml.MergeOptionsListAppend}
+	ret, err := merge2.MergeStrings(string(patchYAML), string(sourceYAML), false, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply aseet %s: %v", patchAssetName, err)
+	}
+	return []byte(ret), nil
+}
+
+// Add a sidecar to the source YAML, possibly with its kube-rbac-proxy when the sidecar reports metrics.
+func (gen *AssetGenerator) addSidecar(sourceYAML []byte, sidecarAssetName string, extraReplacements []string, extraArguments []string, flavour ClusterFlavour, assetPatches AssetPatches) ([]byte, error) {
+	sidecarYAML := gen.mustReadAsset(sidecarAssetName, extraReplacements)
+
+	sidecarYAML, err := gen.addArguments(sidecarYAML, extraArguments)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply all assetPatches
+	for _, patch := range assetPatches {
+		if !patch.ClusterFlavours.Has(flavour) {
+			continue
+		}
+		sidecarYAML, err = gen.applyAssetPatch(sidecarYAML, patch.PatchAssetName, extraReplacements)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Apply the sidecar as strategic merge patch against the source YAML. This add the sidecar as a new container
+	// into the source deployment YAML.
+	opts := yaml.MergeOptions{ListIncreaseDirection: yaml.MergeOptionsListAppend}
+	ret, err := merge2.MergeStrings(string(sidecarYAML), string(sourceYAML), false, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply asset %s: %v", sidecarAssetName, err)
```

Comment:
`%v` looks as a typo. Should we use `%w` everywhere, or there is a reason to prefer `%v` here?

Reply:
Fixed.